jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/CachedJMSConnectionFactory.java

```diff
@@ -0,0 +1,277 @@
+import org.wso2.carbon.transport.jms.utils.JMSConstants;
+
+import java.util.Properties;
+
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageProducer;
+import javax.jms.Session;
+
+/**
+ * {@link JMSConnectionFactory} connector, session, consumer, producer caching implementation. This class can be used
+ * if needed to cache connections.
+ */
+public class CachedJMSConnectionFactory extends JMSConnectionFactory {
+    private static final Log logger = LogFactory.getLog(CachedJMSConnectionFactory.class);
+
+    private int cacheLevel = 0;
+
+    private Connection cachedConnection = null;
+    private Session cachedSession = null;
+    private MessageConsumer cachedMessageConsumer = null;
+    private MessageProducer cachedMessageProducer = null;
+
+    /**
+     * Constructor to create CachedJMSConnectionFactory instance.
+     *
+     * @param properties Property values required to initialize the connection factory
+     * @throws JMSConnectorException
+     */
+    public CachedJMSConnectionFactory(Properties properties) throws JMSConnectorException {
+        super(properties);
+        setValues(properties);
+    }
+
+    private void setValues(Properties properties) {
+        String cacheLevel = properties.getProperty(JMSConstants.PARAM_CACHE_LEVEL);
+        if (null != cacheLevel && !"".equals(cacheLevel)) {
+            this.cacheLevel = Integer.parseInt(cacheLevel);
+        } else {
+            this.cacheLevel = JMSConstants.CACHE_NONE;
+        }
+    }
+
+    @Override public ConnectionFactory getConnectionFactory() throws JMSConnectorException {
+        return super.getConnectionFactory();
+    }
+
+    @Override public Connection getConnection() throws JMSException {
+        return getConnection(null, null);
+    }
+
+    @Override public Connection getConnection(String userName, String password) throws JMSException {
+        Connection connection;
+        if (cachedConnection == null) {
+            connection = createConnection(userName, password);
+        } else {
+            connection = cachedConnection;
+        }
+        if (connection == null) {
+            return null;
+        }
+        try {
+            connection.start();
+        } catch (JMSException e) {
+            if (cachedConnection != null) {
+                resetCache();
+                getConnection(userName, password);
+            } else {
+                logger.error(
+                        "JMS Exception while starting connection for factory '" + this.connectionFactoryString + "' " +
+                        e.getMessage());
+                throw e;
+            }
+
+        }
+        return connection;
+    }
+
+    @Override public Connection createConnection(String userName, String password) throws JMSException {
+        Connection connection;
+        if (userName == null || password == null) {
```

Comment:
That will cause connector caching and cached connector loading logic to be duplicated.